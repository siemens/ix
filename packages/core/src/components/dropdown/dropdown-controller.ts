/*
 * SPDX-FileCopyrightText: 2024 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { IxComponent } from '../utils/internal';
export type CloseBehavior = 'inside' | 'outside' | 'both' | boolean;

export interface DropdownInterface extends IxComponent {
  closeBehavior: CloseBehavior;
  discoverAllSubmenus: boolean;

  getAssignedSubmenuIds(): string[];
  getId(): string;

  discoverSubmenu(): void;

  isPresent(): boolean;

  willPresent?(): boolean;
  willDismiss?(): boolean;

  present(): void;
  dismiss(): void;
}

export function hasDropdownItemWrapperImplemented(
  item: unknown
): item is DropdownItemWrapper {
  return (
    item &&
    (item as DropdownItemWrapper).getDropdownItemElement !== undefined &&
    typeof (item as DropdownItemWrapper).getDropdownItemElement === 'function'
  );
}

export interface DropdownItemWrapper {
  getDropdownItemElement(): Promise<HTMLIxDropdownItemElement>;
}

type SubmenuIds = Record<string, string[]>;

class DropdownController {
  private dropdowns: Map<string, DropdownInterface> = new Map<
    string,
    DropdownInterface
  >();
  private submenuIds: SubmenuIds = {};

  private isWindowListenerActive = false;

  connected(dropdown: DropdownInterface) {
    if (!this.isWindowListenerActive) {
      this.addOverlayListeners();
    }
    this.dropdowns.set(dropdown.getId(), dropdown);

    if (dropdown.discoverAllSubmenus) {
      this.discoverSubmenus();
    }
  }

  disconnected(dropdown: DropdownInterface) {
    this.dropdowns.delete(dropdown.getId());
  }

  discoverSubmenus() {
    this.dropdowns.forEach((dropdown) => {
      dropdown.discoverSubmenu();
    });
  }

  present(dropdown: DropdownInterface) {
    if (!dropdown.isPresent() && dropdown.willPresent()) {
      this.submenuIds[dropdown.getId()] = dropdown.getAssignedSubmenuIds();
      dropdown.present();
    }
  }

  dismissChildren(uid: string) {
    const childIds = this.submenuIds[uid] || [];
    for (const id of childIds) {
      this.dismiss(this.dropdowns.get(id));
    }
  }

  dismiss(dropdown: DropdownInterface) {
    if (dropdown.isPresent() && dropdown.willDismiss()) {
      this.dismissChildren(dropdown.getId());
      dropdown.dismiss();
      delete this.submenuIds[dropdown.getId()];
    }
  }

  dismissAll(
    ignoreBehaviorForIds: string[] = [],
    ignoreRelatedDropdowns = false
  ) {
    this.dropdowns.forEach((dropdown) => {
      const preventClosing =
        dropdown.closeBehavior === 'inside' || dropdown.closeBehavior === false;

      const shouldIgnore = ignoreBehaviorForIds.includes(dropdown.getId());
      const path = this.buildComposedPath(dropdown.getId(), new Set<string>());

      if (ignoreBehaviorForIds.length > 0 && ignoreRelatedDropdowns) {
        let skipRelatedDropdown = false;

        ignoreBehaviorForIds.forEach((id) => {
          if (path.has(id)) {
            skipRelatedDropdown = true;
            return;
          }
        });

        if (!skipRelatedDropdown) {
          return;
        }
      }

      if (!shouldIgnore && preventClosing) {
        return;
      }

      this.dismiss(dropdown);
    });
  }

  dismissOthers(uid: string) {
    let path = this.buildComposedPath(uid, new Set<string>());

    this.dropdowns.forEach((dropdown) => {
      if (
        dropdown.closeBehavior !== 'inside' &&
        dropdown.closeBehavior !== false &&
        !path.has(dropdown.getId())
      ) {
        this.dismiss(dropdown);
      }
    });
  }

  pathIncludesTrigger(eventTargets: EventTarget[]) {
    for (let eventTarget of eventTargets) {
      if (eventTarget instanceof HTMLElement) {
        if (eventTarget.hasAttribute('data-ix-dropdown-trigger')) {
          return eventTarget;
        }
      }
    }

    return;
  }

  private pathIncludesDropdown(eventTargets: EventTarget[]) {
    return !!eventTargets.find(
      (element: HTMLElement) => element.tagName === 'IX-DROPDOWN'
    );
  }

  private buildComposedPath(id: string, path: Set<string>): Set<string> {
    if (this.submenuIds[id]) {
      path.add(id);
    }

    for (const ruleKey of Object.keys(this.submenuIds)) {
      if (this.submenuIds[ruleKey].includes(id)) {
        this.buildComposedPath(ruleKey, path).forEach((key) => path.add(key));
      }
    }

    return path;
  }

  private addOverlayListeners() {
    this.isWindowListenerActive = true;

    window.addEventListener('click', (event: PointerEvent) => {
      const hasTrigger = this.pathIncludesTrigger(event.composedPath());
      const hasDropdown = this.pathIncludesDropdown(event.composedPath());

      if (!hasTrigger && !hasDropdown) {
        this.dismissAll();
      }
    });

    window.addEventListener('keydown', (event: KeyboardEvent) => {
      if (event.key === 'Escape') {
        this.dismissAll([...this.dropdowns.keys()]);
      }
    });
  }
}

export const addDisposableEventListener = (
  element: Element | Window | Document,
  eventType: string,
  callback: EventListenerOrEventListenerObject
) => {
  element.addEventListener(eventType, callback);

  return () => {
    element.removeEventListener(eventType, callback);
  };
};

export const addDisposableEventListenerAsArray = (
  listener: {
    element: Element | Window | Document;
    eventType: string;
    callback: EventListenerOrEventListenerObject;
  }[]
) => {
  const disposables = listener.map(({ callback, element, eventType }) =>
    addDisposableEventListener(element, eventType, callback)
  );

  return () => disposables.forEach((dispose) => dispose());
};

export const dropdownController = new DropdownController();
