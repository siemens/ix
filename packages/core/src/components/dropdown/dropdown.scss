/*
 * SPDX-FileCopyrightText: 2023 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

@import 'common-variables';

@import 'mixins/fonts';
@import 'mixins/text-truncation';
@import 'mixins/shadow-dom/component';
@import 'components/dropdown';

:host {
  @include ix-component;
  min-width: 0px;
  z-index: var(--theme-z-index-dropdown);
}

:host(.overflow) {
  max-height: 50vh;
  overflow-y: auto;
}

:host(:not(.show)) {
  display: none;
}
