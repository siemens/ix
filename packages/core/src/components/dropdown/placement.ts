/*
 * SPDX-FileCopyrightText: 2022 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
export declare const top: 'top';
export declare const bottom: 'bottom';
export declare const right: 'right';
export declare const left: 'left';

export declare type BasePlacement =
  | typeof top
  | typeof bottom
  | typeof right
  | typeof left;

export declare type AutoPlacement = 'auto' | 'auto-start' | 'auto-end';

export declare type PlacementWithAlignment =
  | 'top-start'
  | 'top-end'
  | 'bottom-start'
  | 'bottom-end'
  | 'right-start'
  | 'right-end'
  | 'left-start'
  | 'left-end';

export declare type Placement =
  | AutoPlacement
  | BasePlacement
  | PlacementWithAlignment;
