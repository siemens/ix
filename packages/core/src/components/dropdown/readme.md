<!--
SPDX-FileCopyrightText: 2022 Siemens AG

SPDX-License-Identifier: MIT
-->

# ix-dropdown

<!-- Auto Generated Below -->


## Properties

| Property                              | Attribute                                  | Description                                                                                                                                                                                         | Type                                                                                                                                                                                                         | Default        |
| ------------------------------------- | ------------------------------------------ | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ | -------------- |
| `adjustDropdownWidthToReferenceWidth` | `adjust-dropdown-width-to-reference-width` | Adjust dropdown width to the parent width                                                                                                                                                           | `boolean`                                                                                                                                                                                                    | `false`        |
| `adjustDropdownWidthToReferenceWith`  | `adjust-dropdown-width-to-reference-with`  | <span style="color:red">**[DEPRECATED]**</span> Will be removed in 2.0.0. Property has a typo use `adjustDropdownWidthToReferenceWidth` instead.<br/><br/>Adjust dropdown width to the parent width | `boolean`                                                                                                                                                                                                    | `false`        |
| `anchor`                              | `anchor`                                   | Define an anchor element                                                                                                                                                                            | `HTMLElement \| string`                                                                                                                                                                                      | `undefined`    |
| `closeBehavior`                       | `close-behavior`                           | Close behavior                                                                                                                                                                                      | `"both" \| "inside" \| "outside" \| boolean`                                                                                                                                                                 | `'both'`       |
| `header`                              | `header`                                   | An optional header shown at the top of the dropdown                                                                                                                                                 | `string`                                                                                                                                                                                                     | `undefined`    |
| `placement`                           | `placement`                                | Placement of the dropdown                                                                                                                                                                           | `"auto" \| "auto-end" \| "auto-start" \| "bottom" \| "bottom-end" \| "bottom-start" \| "left" \| "left-end" \| "left-start" \| "right" \| "right-end" \| "right-start" \| "top" \| "top-end" \| "top-start"` | `'bottom-end'` |
| `positioningStrategy`                 | `positioning-strategy`                     | Position strategy                                                                                                                                                                                   | `"absolute" \| "fixed"`                                                                                                                                                                                      | `'fixed'`      |
| `show`                                | `show`                                     | Show dropdown                                                                                                                                                                                       | `boolean`                                                                                                                                                                                                    | `false`        |
| `trigger`                             | `trigger`                                  | Define an element that triggers the dropdown. A trigger can either be a string that will be interprated as id attribute or a DOM element.                                                           | `HTMLElement \| string`                                                                                                                                                                                      | `undefined`    |


## Events

| Event         | Description                                         | Type                   |
| ------------- | --------------------------------------------------- | ---------------------- |
| `showChanged` | Fire event after visibility of dropdown has changed | `CustomEvent<boolean>` |


## Methods

### `updatePosition() => Promise<void>`

Update position of dropdown

#### Returns

Type: `Promise<void>`




----------------------------------------------


