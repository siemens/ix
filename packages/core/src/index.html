<!--
SPDX-FileCopyrightText: 2023 Siemens AG

SPDX-License-Identifier: MIT
-->

<!DOCTYPE html>
<html dir="ltr" lang="en">

<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0, minimum-scale=1.0, maximum-scale=5.0" />
  <title>Stencil Component Starter</title>

  <script type="module" src="/build/siemens-ix.esm.js"></script>
  <script nomodule src="/build/siemens-ix.js"></script>

  <link rel="stylesheet" href="/build/siemens-ix.css" />
  <link rel="stylesheet" href="/build/ix-brand-theme/dist/ix-brand-theme/ix-brand-theme.css" />
  <link rel="stylesheet" href="/build/ix-icons/css/ix-icons.css" />
  <script type="module">
    import { defineCustomElements } from '/build/ix-brand-theme/loader/index.es2017.js';
    defineCustomElements();
  </script>
</head>

<body style="margin: 0px; height: 100vh">
  <my-component />
</body>

</html>