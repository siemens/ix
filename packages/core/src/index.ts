/*
 * SPDX-FileCopyrightText: 2023 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import '@siemens/ix-icons';
import type {
  Components as IxIconsComponents,
  JSX as IxIconsJSX,
} from '@siemens/ix-icons';
export * from './components';
export * from './components/category-filter/filter-state';
export * from './components/category-filter/input-state';
export * from './components/category-filter/logical-filter-operator';
export { FlipTileState } from './components/flip-tile/flip-tile-state';
export * from './components/modal/modal-utils';
export * from './components/toast/toast-utils';
export * from './components/tree-item/default-tree-item';
export * from './components/tree/tree-model';
export * from './components/upload/upload-file-state';
export { NotificationColor } from './components/utils/notification-color';
export {
  convertToAbbreviationString,
  convertToRemString,
} from './components/utils/rwd.util';
export * from './components/utils/theme-switcher';

declare module './components' {
  export namespace Components {
    export type IxIcon = IxIconsComponents.IxIcon;
  }
}

declare module './components' {
  export namespace JSX {
    export type IxIcon = IxIconsJSX.IxIcon;
  }
}
