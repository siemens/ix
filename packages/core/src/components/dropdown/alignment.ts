/*
 * SPDX-FileCopyrightText: 2022 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { Placement } from './placement';

export function getAlignment(placement: Placement) {
  if (placement.includes('-end')) {
    return 'end';
  } else if (placement.includes('-start')) {
    return 'start';
  }

  return undefined;
}
