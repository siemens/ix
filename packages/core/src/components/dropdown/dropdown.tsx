/*
 * SPDX-FileCopyrightText: 2023 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {
  autoUpdate,
  computePosition,
  ComputePositionConfig,
  flip,
  inline,
  offset,
  shift,
} from '@floating-ui/dom';
import {
  Component,
  Element,
  Event,
  EventEmitter,
  h,
  Host,
  Listen,
  Method,
  Prop,
  Watch,
} from '@stencil/core';
import { BasePlacement, Placement, PlacementWithAlignment } from './placement';

/**
 * @internal
 */
export type DropdownTriggerEvent = 'click' | 'hover' | 'focus';

@Component({
  tag: 'ix-dropdown',
  styleUrl: 'dropdown.scss',
  shadow: true,
})
export class Dropdown {
  @Element() hostElement!: HTMLIxDropdownElement;

  /**
   * Show dropdown
   */
  @Prop({ mutable: true, reflect: true }) show = false;

  /**
   * Define an element that triggers the dropdown.
   * A trigger can either be a string that will be interpreted as id attribute or a DOM element.
   */
  @Prop() trigger: string | HTMLElement;

  /**
   * Define an anchor element
   */
  @Prop() anchor: string | HTMLElement;

  /**
   * Close behavior
   */
  @Prop() closeBehavior: 'inside' | 'outside' | 'both' | boolean = 'both';

  /**
   * Placement of the dropdown
   */
  @Prop() placement: Placement = 'bottom-start';

  /**
   * Position strategy
   */
  @Prop() positioningStrategy: 'absolute' | 'fixed' = 'fixed';

  /**
   * Adjust dropdown width to the parent width
   * @deprecated Will be removed in 2.0.0. Property has a typo use `adjustDropdownWidthToReferenceWidth` instead.
   */
  @Prop() adjustDropdownWidthToReferenceWith = false;

  /**
   * Adjust dropdown width to the parent width
   *
   * @deprecated Will be removed. Not used anymore
   */
  @Prop() adjustDropdownWidthToReferenceWidth = false;

  /**
   * An optional header shown at the top of the dropdown
   */
  @Prop() header?: string;

  /**
   * Move dropdown along main axis of alignment
   *
   * @internal
   */
  @Prop() offset: {
    mainAxis?: number;
    crossAxis?: number;
    alignmentAxis?: number;
  };

  /**
   * Define one or more events to open dropdown
   * @internal
   */
  @Prop() triggerEvent: DropdownTriggerEvent | DropdownTriggerEvent[] = 'click';

  /**
   * @internal
   */
  @Prop() overwriteDropdownStyle: (delegate: {
    dropdownRef: HTMLElement;
    triggerRef?: HTMLElement;
  }) => Promise<Partial<CSSStyleDeclaration>>;

  /**
   * Fire event after visibility of dropdown has changed
   */
  @Event() showChanged: EventEmitter<boolean>;

  private autoUpdateCleanup: () => void = null;

  private triggerElement?: Element;
  private anchorElement?: Element;

  private dropdownRef: HTMLElement;

  private toggleBind: any;
  private openBind: any;

  constructor() {
    this.toggleBind = this.toggle.bind(this);
    this.openBind = this.open.bind(this);
  }

  get dropdownItems() {
    return Array.from(this.hostElement.querySelectorAll('ix-dropdown-item'));
  }

  private addEventListenersFor(triggerEvent: DropdownTriggerEvent) {
    switch (triggerEvent) {
      case 'click':
        if (this.closeBehavior === 'outside') {
          this.triggerElement.addEventListener('click', this.openBind);
        } else {
          this.triggerElement.addEventListener('click', this.toggleBind);
        }
        break;

      case 'hover':
        this.triggerElement.addEventListener('mouseenter', this.openBind);
        break;

      case 'focus':
        this.triggerElement.addEventListener('focusin', this.openBind);
        break;
    }
  }

  private removeEventListenersFor(
    triggerEvent: DropdownTriggerEvent,
    triggerElement: Element
  ) {
    switch (triggerEvent) {
      case 'click':
        if (this.closeBehavior === 'outside') {
          triggerElement.removeEventListener('click', this.openBind);
        } else {
          triggerElement.removeEventListener('click', this.toggleBind);
        }
        break;

      case 'hover':
        triggerElement.removeEventListener('mouseenter', this.openBind);
        break;

      case 'focus':
        triggerElement.removeEventListener('focusin', this.openBind);
        break;
    }
  }

  private async registerListener(element: string | HTMLElement) {
    this.triggerElement = await this.resolveElement(element);
    if (this.triggerElement) {
      if (Array.isArray(this.triggerEvent)) {
        this.triggerEvent.forEach((triggerEvent) => {
          this.addEventListenersFor(triggerEvent);
        });
      } else {
        this.addEventListenersFor(this.triggerEvent);
      }
    }
  }

  private async unregisterListener(element: string | HTMLElement) {
    const trigger = await this.resolveElement(element);
    if (Array.isArray(this.triggerEvent)) {
      this.triggerEvent.forEach((triggerEvent) => {
        this.removeEventListenersFor(triggerEvent, trigger);
      });
    } else {
      this.removeEventListenersFor(this.triggerEvent, trigger);
    }
  }

  private resolveElement(element: string | HTMLElement): Promise<Element> {
    if (typeof element !== 'string') {
      return Promise.resolve(element);
    }

    const selector = `#${element}`;
    return new Promise((resolve) => {
      if (document.querySelector(selector)) {
        return resolve(document.querySelector(selector));
      }

      const observer = new MutationObserver(() => {
        if (document.querySelector(selector)) {
          resolve(document.querySelector(selector));
          observer.disconnect();
        }
      });

      observer.observe(document.body, {
        childList: true,
        subtree: true,
      });
    });
  }

  @Watch('show')
  async changedShow(newShow: boolean) {
    if (newShow) {
      this.anchorElement = await (this.anchor
        ? this.resolveElement(this.anchor)
        : this.resolveElement(this.trigger));

      if (this.anchorElement) {
        this.applyDropdownPosition();
      }
    }
  }

  @Watch('trigger')
  changedTrigger(
    newTriggerValue: string | HTMLElement,
    oldTriggerValue: string | HTMLElement
  ) {
    if (newTriggerValue) {
      this.registerListener(newTriggerValue);
    }

    if (oldTriggerValue) {
      this.unregisterListener(oldTriggerValue);
    }
  }

  @Listen('click', {
    target: 'window',
  })
  clickOutside(event: Event) {
    const target = event.target as HTMLElement;

    if (
      this.show === false ||
      this.closeBehavior === false ||
      this.anchorElement?.contains(target) ||
      this.triggerElement?.contains(target)
    ) {
      return;
    }

    switch (this.closeBehavior) {
      case 'outside':
        if (!this.dropdownRef.contains(target)) {
          this.close();
        }
        break;
      case 'inside':
        if (this.dropdownRef.contains(target) && this.hostElement !== target) {
          this.close();
        }
        break;
      case 'both':
        if (this.hostElement !== target) this.close();
        break;
      default:
        this.close();
    }
  }

  private isNestedDropdown(element: Element) {
    return element.closest('ix-dropdown');
  }

  private isAnchorSubmenu() {
    const anchor = this.anchorElement?.closest('ix-dropdown-item');
    if (!anchor) {
      return false;
    }

    return true;
  }

  private toggle(event?: Event) {
    event?.preventDefault();

    if (this.isNestedDropdown(event.target as HTMLElement)) {
      event?.stopPropagation();
    }

    this.show = !this.show;
    this.showChanged.emit(this.show);
  }

  private open(event?: Event) {
    event?.preventDefault();

    if (this.isNestedDropdown(event.target as HTMLElement)) {
      event?.stopPropagation();
    }

    this.show = true;
    this.showChanged.emit(true);
  }

  private close(event?: Event) {
    if (event?.defaultPrevented) {
      return;
    }

    this.show = false;
    this.showChanged.emit(false);
  }

  private async applyDropdownPosition() {
    if (!this.anchorElement) {
      return;
    }
    if (!this.dropdownRef) {
      return;
    }
    const isSubmenu = this.isAnchorSubmenu();

    let positionConfig: Partial<ComputePositionConfig> = {
      strategy: this.positioningStrategy,
      middleware: [],
    };

    if (isSubmenu) {
      positionConfig.placement = 'right-start';
    }

    if (this.placement.includes('auto') || isSubmenu) {
      positionConfig.middleware.push(
        flip({ fallbackStrategy: 'initialPlacement' })
      );
      positionConfig.placement = 'bottom-start';
    } else {
      positionConfig.placement = this.placement as
        | BasePlacement
        | PlacementWithAlignment;
    }

    positionConfig.middleware = [
      ...positionConfig.middleware,
      inline(),
      shift(),
    ];

    if (this.offset) {
      positionConfig.middleware.push(offset(this.offset));
    }

    if (this.autoUpdateCleanup) {
      this.autoUpdateCleanup();
      this.autoUpdateCleanup = null;
    }
    this.autoUpdateCleanup = autoUpdate(
      this.anchorElement,
      this.dropdownRef,
      async () => {
        const computeResponse = await computePosition(
          this.anchorElement,
          this.dropdownRef,
          positionConfig
        );
        Object.assign(this.dropdownRef.style, {
          top: '0',
          left: '0',
          transform: `translate(${Math.round(computeResponse.x)}px,${Math.round(
            computeResponse.y
          )}px)`,
        });
        if (this.overwriteDropdownStyle) {
          const overwriteStyle = await this.overwriteDropdownStyle({
            dropdownRef: this.dropdownRef,
            triggerRef: this.triggerElement as HTMLElement,
          });

          Object.assign(this.dropdownRef.style, overwriteStyle);
        }
      },
      {
        ancestorResize: true,
        ancestorScroll: true,
        elementResize: true,
      }
    );
  }

  async componentDidLoad() {
    if (this.trigger) {
      this.registerListener(this.trigger);
    }
  }

  async componentDidRender() {
    await this.applyDropdownPosition();
  }

  disconnectedCallback() {
    if (this.autoUpdateCleanup) {
      this.autoUpdateCleanup();
    }
  }

  /**
   * Update position of dropdown
   */
  @Method()
  async updatePosition() {
    this.applyDropdownPosition();
  }

  render() {
    return (
      <Host
        ref={(ref) => (this.dropdownRef = ref)}
        class={{
          'dropdown-menu': true,
          show: this.show,
          overflow: true,
        }}
        style={{
          margin: '0',
          minWidth: '0px',
          position: this.positioningStrategy,
        }}
      >
        <div style={{ display: 'contents' }}>
          {this.header ? <div class="dropdown-header">{this.header}</div> : ''}

          <slot></slot>
        </div>
      </Host>
    );
  }
}
