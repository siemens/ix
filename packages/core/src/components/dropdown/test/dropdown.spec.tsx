/*
 * SPDX-FileCopyrightText: 2023 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { newSpecPage } from '@stencil/core/testing';
import { fireEvent } from '@testing-library/dom';
import { Dropdown } from '../dropdown';

describe('ix-dropdown', () => {
  let page: any;

  it('should open with anchor element', async () => {
    const page = await newSpecPage({
      components: [Dropdown],
      html: '<div></div>',
    });

    const anchor = page.doc.createElement('ix-button');
    anchor.setAttribute('id', 'test');
    anchor.innerText = 'Test';

    const dropdown = page.doc.createElement('ix-dropdown');
    dropdown.show = true;

    const dropdownItem = page.doc.createElement('ix-dropdown-item');

    dropdown.appendChild(dropdownItem);

    page.root.appendChild(anchor);
    page.root.appendChild(dropdown);

    await page.waitForChanges();

    expect(page.root).toEqualHtml(`
    <ix-dropdown class="dropdown-menu overflow show" role="list" show="" style="margin: 0; min-width: 0px; position: fixed;">
      <mock:shadow-root>
        <div style="display: contents;">
          <slot></slot>
        </div>
      </mock:shadow-root>
      <ix-dropdown-item></ix-dropdown-item>
    </ix-dropdown>
    `);
  });

  it('should collapse, when clicked outside', async () => {
    const page = await newSpecPage({
      components: [Dropdown],
      html: `<div></div>`,
    });

    const dropdown = page.doc.createElement('ix-dropdown');
    dropdown.show = true;
    page.root.appendChild(dropdown);

    await page.waitForChanges();

    fireEvent.click(window);
    await page.waitForChanges();

    expect(dropdown.show).toBeFalsy();
  });

  it('emits an event, when show changed', async () => {
    const mockCallback = jest.fn();
    const page = await newSpecPage({
      components: [Dropdown],
      html: `<div></div>`,
    });
    page.win.addEventListener('showChanged', mockCallback);

    const dropdown = page.doc.createElement('ix-dropdown');
    dropdown.show = true;
    page.root.appendChild(dropdown);
    await page.waitForChanges();

    fireEvent.click(window);
    await page.waitForChanges();
    expect(mockCallback).toHaveBeenCalled();
  });
});
