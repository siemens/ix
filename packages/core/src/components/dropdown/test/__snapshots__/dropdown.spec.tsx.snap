// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ix-dropdown renders 1`] = `
<ix-dropdown class="dropdown-menu overflow" style="margin: 0; min-width: 0px; position: fixed;">
  <mock:shadow-root>
    <div style="display: contents;">
      <slot></slot>
    </div>
  </mock:shadow-root>
  <ix-dropdown-item></ix-dropdown-item>
</ix-dropdown>
`;
